### PR TITLE
Core/ts version task 357 refactor swap executer

### PR DESCRIPTION
**Pull request by franken5tein**
## Check list  

- [x] All Merge Request pipelines should be passed.
- [x] Make sure all the changes are necessary and correct.
- [x] Make sure no needed change missing (all necessary changes are made).
- [x] Merge Request should contains only changes for one purpose (one feature or bug or refactoring).
- [ ] All necessary unit tests are written.
- [x] Critical changes should be tested locally (or on staging).

### CODE REVIEW

File: ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts [hunk 1]
```diff
@@ -111,21 +111,20 @@ class ZkSyncV1 implements IBlockchain {
...
 
-    network = network as Network;
-    market = market as Market;
-    makerOrder = makerOrder as Order;
-    takerOrder = takerOrder as Order;
-
-    const baseToken = market.info.base_asset;
-    const quoteToken = market.info.quote_asset;
-    const broadcaster = await this._getBroadcaster();
+    const baseToken = Network.getNetwork(_network as string)?.getMarket(_market)
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts:118]: use fill.market and fill.network

File: ts-version/src/data/blockchain/evm.ts [hunk 2]
```diff
@@ -387,49 +367,77 @@ export default class Evm implements IBlockchain {
...
+      instance.start_promise = instance.start();
 
       this._instances.set(name, instance);
     }
 
     return this._instances.get(name) as Evm;
   }
 
-  private async _getBroadcaster(): Promise<broadcaster | void> {
+  public async getBroadcaster(): Promise<void | broadcaster | undefined> {
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/evm.ts:392]: getBroadcaster and getBroadcasters functions have same body. as i see, getBroadcasters usecase is just returning all broadcasters.

File: ts-version/src/data/blockchain/evm.ts [hunk 3]
```diff
@@ -387,49 +369,51 @@ export default class Evm implements IBlockchain {
...
-            // check if account has enough ether to broadcast swaps!
-            selectedWallet = this.broadcasters[wallet as unknown as number];
-            selectedWallet.orderBroadcasting = true;
-
-            return selectedWallet;
-          }
-        }
-
-        await new Promise(resolve => setTimeout(resolve, 2000));
+  public async getBroadcasters(): Promise<broadcaster[] | undefined> {
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/evm.ts:394]: i think this function should just return this.broadcaster

File: ts-version/src/utils/config/swap-executers.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    });
+  }
+
+  public async prepareFill(fill: Fill): Promise<void> {
+    const swapExecuter = this.get(fill._network as string);
+    await swapExecuter.prepareFill({ fill });
+  }
+
+  public static getInstance(
+    orderRepo: OrderRepoAdapter,
```
Comment (hypersolo) [inline ts-version/src/utils/config/swap-executers.ts:78]: getInstance parameters must be optional

File: ts-version/src/domain/services/swap-executer/swap-executer.ts [hunk 5]
```diff
@@ -3,76 +3,198 @@ import { IFillRepo } from "@port/repository/fill";
...
+
+    broadcaster.orderBroadcasting = true;
+    SwapExecuter.broadcasters.set(network, broadcasters);
 
-    let matchedOrder: prepareFillOutput = [];
+    return broadcaster;
+  }
+
+  private _releaseBroadcaster(network: string, broadcaster: broadcaster): void {
+    const broadcasters = SwapExecuter.broadcasters.get(network);
```
Comment (hypersolo) [inline ts-version/src/domain/services/swap-executer/swap-executer.ts:63]: doesn't broadcaster.orderBroadcasting work?

File: ts-version/src/domain/services/swap-executer/swap-executer.ts [hunk 6]
```diff
@@ -3,76 +3,198 @@ import { IFillRepo } from "@port/repository/fill";
...
+    if (broadcasters)
+      broadcasters.map(b => {
+        if (b === broadcaster) b.orderBroadcasting = false;
+      });
+    SwapExecuter.broadcasters.set(network, broadcasters as broadcaster[]);
+  }
+
+  public async prepareFill(
```
Comment (hypersolo) [inline ts-version/src/domain/services/swap-executer/swap-executer.ts:71]: processFill might be better function name.

File: ts-version/src/domain/services/swap-executer/swap-executer.ts [hunk 7]
```diff
@@ -3,76 +3,198 @@ import { IFillRepo } from "@port/repository/fill";
...
+    data: ISwapExecuterService.executeSwapInput,
+  ): Promise<executeSwapOutput | void> {
+    const { fill } = data;
+
+    let matchedOrder: matchedOrder | undefined;
+    const broadcaster = this._getBroadcaster(fill._network as string);
```
Comment (hypersolo) [inline ts-version/src/domain/services/swap-executer/swap-executer.ts:77]: this getter function locks the broadcaster too. but here we are just srializing fill, and broadcaster is not busy. it's better to move locking and unlocking logic to executeSwaps function.

Comment (hypersolo) [inline ts-version/src/domain/services/swap-executer/swap-executer.ts:77]: this getter function locks the broadcaster too. but here we are just srializing fill, and broadcaster is not busy. it's better to move locking and unlocking logic to executeSwap function of IBlockchain classes. so we need to remove locking logic from _getBroadcaster function.

File: ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts (hunk 1 above)
Comment (franken5tein) [reply ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts]: Done!

File: ts-version/src/data/blockchain/evm.ts [hunk 8]
```diff
@@ -184,27 +181,25 @@ export default class Evm implements IBlockchain {
...
     } = fill;
 
-    network = network as Network;
-    market = market as Market;
     makerOrder = makerOrder as Order;
     takerOrder = takerOrder as Order;
 
-    const baseDecimal = Network.getNetwork(network.name)
-      ?.markets.get(market.name)
+    const baseDecimal = Network.getNetwork(_network as string)
```
[the remark this answers is not in the document]
Comment (franken5tein) [reply ts-version/src/data/blockchain/evm.ts]: Done!

File: ts-version/src/app.ts [hunk 9]
```diff
@@ -79,17 +81,23 @@ async function main() {
...
   schedule(jobPeriods.processStarvedOrders, () => {
     processStarvedOrders.process();
   });
-  const priceService = PriceService.getInstance(priceAPI);
-  PriceAPI.register("stats_updated", priceService as unknown as IEventObserver);
   schedule(jobPeriods.contractEvents, async () => {
     await Events.getInstance(networkRepo, fillRepo, orderRepo).startEvents();
   });
+  schedule(jobPeriods.swapExecuter, async () => {
+    await swapExecuters.startSchedules();
```
[the remark this answers is not in the document]
Comment (franken5tein) [reply ts-version/src/app.ts]: Done!

File: ts-version/src/domain/services/swap-executer/swap-executer.ts (hunk 6 above)
Comment (franken5tein) [reply ts-version/src/domain/services/swap-executer/swap-executer.ts]: Done! https://github.com/dexpresso-protocol/core/commit/551c6d5e8aec30aedd2eacbff231527cd38d764d

File: ts-version/src/utils/config/swap-executers.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    );
+  }
+
+  public get(name: string): ISwapExecuterService.schema {
+    return this.swapExecuters.get(name) as ISwapExecuterService.schema;
+  }
+
+  public async startSchedules(): Promise<void> {
+    this.swapExecuters.forEach(async swapExecuter => {
+      await swapExecuter.executeQueuedSwaps();
```
[the remark this answers is not in the document]
Comment (franken5tein) [reply ts-version/src/utils/config/swap-executers.ts]: Done! https://github.com/dexpresso-protocol/core/commit/551c6d5e8aec30aedd2eacbff231527cd38d764d

File: ts-version/src/domain/services/swap-executer/swap-executer.ts (hunk 5 above)
Comment (franken5tein) [reply ts-version/src/domain/services/swap-executer/swap-executer.ts]: Done! https://github.com/dexpresso-protocol/core/commit/b3ac88ab6cd3fb40aa4ecaf13328474dc43ce9dc

File: ts-version/src/domain/services/swap-executer/swap-executer.ts [hunk 11]
```diff
@@ -3,76 +3,198 @@ import { IFillRepo } from "@port/repository/fill";
...
+    if (!broadcasters) return undefined;
 
-    if (status !== FillStatus.match) {
-      return;
-    }
+    const broadcaster = broadcasters.find(b => !b.orderBroadcasting);
+    if (!broadcaster) return undefined;
+
+    broadcaster.orderBroadcasting = true;
+    SwapExecuter.broadcasters.set(network, broadcasters);
```
[the remark this answers is not in the document]
Comment (franken5tein) [reply ts-version/src/domain/services/swap-executer/swap-executer.ts]: Done! https://github.com/dexpresso-protocol/core/commit/b3ac88ab6cd3fb40aa4ecaf13328474dc43ce9dc

File: ts-version/src/domain/services/swap-executer/swap-executer.ts [hunk 12]
```diff
@@ -3,76 +3,198 @@ import { IFillRepo } from "@port/repository/fill";
...
+      return await this.executeSwaps([data]);
+    } else {
+      this.Queue.enqueue(data);
+    }
+  }
+
+  public async executeSwaps(
+    data: preparedFill[],
+  ): Promise<executeSwapOutput | undefined> {
+    const broadcaster = this._getBroadcaster(data[0].network as string);
```
[the remark this answers is not in the document]
Comment (franken5tein) [reply ts-version/src/domain/services/swap-executer/swap-executer.ts:128]: Done! https://github.com/dexpresso-protocol/core/commit/b3ac88ab6cd3fb40aa4ecaf13328474dc43ce9dc

File: ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts [hunk 13]
```diff
@@ -519,53 +516,32 @@ class ZkSyncV1 implements IBlockchain {
...
-        }
-
-        await new Promise((resolve) => setTimeout(resolve, 2000));
-      }
-    }
-  }
+  public async getBroadcasters(): Promise<broadcaster[] | undefined> {
+    const minimumBalance = 0.002;
+    const math = Utilities.getInstance().math;
+    const broadcasters = await Object.values(this.broadcasters).filter(
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts:522]: delete await

File: ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts [hunk 14]
```diff
@@ -88,36 +86,43 @@ class ZkSyncV1 implements IBlockchain {
...
-
-    const baseToken = market.info.base_asset;
-    const quoteToken = market.info.quote_asset;
-    const broadcaster = await this._getBroadcaster();
-    let batchBuilder;
+    const baseToken = fill.market?.info.base_asset;
+    const quoteToken = fill.market?.info.quote_asset;
+    let batchBuilder: BatchBuilder | undefined;
     if (broadcaster)
       batchBuilder = await this._getBatchBuilder(
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts:124]: delete await

File: ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts [hunk 15]
```diff
@@ -443,7 +442,7 @@ class ZkSyncV1 implements IBlockchain {
       batchBuilder,
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts:442]: delete await

File: ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts [hunk 16]
```diff
@@ -220,14 +217,14 @@ class ZkSyncV1 implements IBlockchain {
...
-    const now = math.div(Date.now(), 1000);
+    const now = Number(math.div(Date.now(), 1000));
 
-    if (math.lt(signedOrder.validFrom!, now))
-      throw new BadRequestError("Invalid valid from");
-    if (math.gt(signedOrder.validUntil, now))
-      throw new BadRequestError("Invalid valid until");
-    if (math.lt(signedOrder.validFrom!, signedOrder.validUntil))
-      throw new BadRequestError("Incompatible valid from and valid until");
+    if (math.gt(signedOrder.validFrom!, now))
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/zksync-v1/zksync-v1.ts:222]: fix lint warning

File: ts-version/src/data/blockchain/evm.ts [hunk 17]
```diff
@@ -371,49 +364,51 @@ export default class Evm implements IBlockchain {
     return status;
   }
 
-  public static getInstance(name: string, networkName: string): Evm {
+  public static getInstance(
```
Comment (hypersolo) [inline ts-version/src/data/blockchain/evm.ts:367]: make parameters optional

File: ts-version/src/domain/services/blockchain-event-handler/blockchain-event-handler.ts [hunk 18]
```diff
@@ -34,13 +34,15 @@ export default class EventHandler
...
     const events = await this.blockchainNetwork.getEvents(
       "*",
-      latestReadBlock ? latestReadBlock + 1 : latestBlockNum,
-      latestBlockNum,
+      latestReadBlock
+        ? (latestReadBlock as number) + 1
+        : (latestBlockNum as number),
+      latestBlockNum as number,
     );
     await this._handler(events as unknown as EVMEvents);
```
Comment (hypersolo) [inline ts-version/src/domain/services/blockchain-event-handler/blockchain-event-handler.ts:42]: delete await

File: ts-version/src/domain/services/swap-executer/swap_listener.ts [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+import Fill from "@entity/fill.entity";
+import { SwapExecuters } from "@util/config/swap-executers";
+import OrderRepoAdapter from "src/data/database/postgres/repository/order.adapter";
+import FillRepoAdapter from "src/data/database/postgres/repository/fill.adapter";
+
+export class SwapListener extends BaseService implements IEventObserver {
+  private static instance: SwapListener;
+
+  private constructor(
+    private OrderAdapter: OrderRepoAdapter,
```
Comment (hypersolo) [inline ts-version/src/domain/services/swap-executer/swap_listener.ts:13]: use interface over RepoAdapter.

idea:
just get adapters in constructor and don't set them into object.
use them to get a SwapExecuters instance and use this.swapExecuters in update function (over SwapExecuters.getInstance(....))

File: ts-version/src/utils/config/providers.ts [hunk 20]
```diff
@@ -36,15 +36,15 @@ export class Providers {
...
       Array.from(this.providers.values()).map(
-        (provider) => provider.startPromise,
+        (provider) => provider.start_promise,
       ),
     );
   }
 
-  getProvider(name: string): IBlockchain | undefined {
-    return this.providers.get(name);
+  public getProvider(name: string): IBlockchain {
```
Comment (hypersolo) [inline ts-version/src/utils/config/providers.ts:47]: previous return type was better.

File: ts-version/src/data/database/postgres/repository/order.adapter.ts [hunk 21]
```diff
@@ -305,7 +305,7 @@ class OrderRepoAdapter extends TransactionalDB implements IOrderRepo.schema {
         result.market,
         result.price,
         order.type,
-        order.signedOrder,
+        result.maker_signed_order,
```
Comment (hypersolo) [inline ts-version/src/data/database/postgres/repository/order.adapter.ts:308]: revert

File: ts-version/src/utils/config/swap-executers.ts [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+import Fill from "@entity/fill.entity";
+
+export class SwapExecuters {
+  private static instance: SwapExecuters;
+  private swapExecuters: Map<string, ISwapExecuter.schema> = new Map<
+    string,
+    ISwapExecuter.schema
+  >();
+
+  private constructor(
```
Comment (hypersolo) [inline ts-version/src/utils/config/swap-executers.ts:18]: idea:
- just use constructor inputs to create objects and don't set them into instance.
- don't pass new ArrayQueue(). do it in SwapExecuter classes if needed.

File: ts-version/src/domain/ports/swap-executer.interface.ts [hunk 23]
```diff
@@ -1,15 +1,17 @@
...
+export namespace ISwapExecuter {
   export type executeSwapInput = {
     fill: Fill;
-    wallet: Address | string;
   };
 
   export interface schema {
-    executeSwap(data: executeSwapInput): Promise<executeSwapOutput | undefined>;
+    setBroadcasters(): Promise<void>;
+    processFill(
```
Comment (hypersolo) [inline ts-version/src/domain/ports/swap-executer.interface.ts:11]: we can simply pass fill: Fill.

File: ts-version/src/utils/config/swap-executers.ts [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+        new ArrayQueue(),
+      ),
+    );
+  }
+
+  public get(name: string): ISwapExecuter.schema {
+    return this.swapExecuters.get(name) as ISwapExecuter.schema;
+  }
+
+  public startSchedules(): void {
```
Comment (hypersolo) [inline ts-version/src/utils/config/swap-executers.ts:66]: better naming.

File: ts-version/src/app.ts [hunk 25]
```diff
@@ -99,12 +106,11 @@ async function main() {
       .process()
       .catch((e) => logger.error(`Starved order processor failed: ${e}`));
   });
-  const priceService = PriceService.getInstance(priceAPI);
-  PriceAPI.register("stats_updated", priceService as unknown as IEventObserver);
   schedule(jobPeriods.contractEvents, async () => {
-    await Events.getInstance(networkRepo, fillRepo, orderRepo)
-      .startEvents()
-      .catch((e) => logger.error(`Event handler failed: ${e}`));
+    await Events.getInstance(networkRepo, fillRepo, orderRepo).startEvents();
```
Comment (hypersolo) [inline ts-version/src/app.ts:110]: bad merge. add catch block.

File: ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+  matchedOrder,
+  preparedFill,
+  transactionResultOutput,
+} from "@type/blockchain.types";
+import Order from "@entity/order/order.entity";
+import Utilities from "src/domain/Utilities";
+import { TransactionReceipt } from "zksync/build/types";
+
+export class ZkSyncV1SwapExecuter implements ISwapExecuter.schema {
+  static queueLimit = 20;
```
Comment (hypersolo) [inline ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts:20]: pointless variables

File: ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+
+  public async setBroadcasters(): Promise<void> {
+    const broadcasters = await this.BlockchainNetwork.getBroadcasters();
+    ZkSyncV1SwapExecuter.broadcasters.set(
+      this.BlockchainNetwork.network,
+      broadcasters as broadcaster[],
+    );
+  }
+
+  private _getBroadcaster(network: string): broadcaster | undefined {
```
Comment (hypersolo) [inline ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts:49]: chack for eth balance

File: ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts (hunk 26 above)
Comment (hypersolo) [inline ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts:20]: delete these

Comment (hypersolo) [inline ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts:20]: delete these

File: ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts (hunk 27 above)
Comment (hypersolo) [inline ts-version/src/data/swap-executer/zksyncv1-swap-executer.ts:49]: - check for balance
- add retry

File: ts-version/src/data/swap-executer/evm-swap-executer.ts [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+  executeSwapOutput,
+  matchedOrder,
+  preparedFill,
+} from "@type/blockchain.types";
+import Order from "@entity/order/order.entity";
+import Utilities from "src/domain/Utilities";
+
+export class EVMSwapExecuter implements ISwapExecuter.schema {
+  static queueLimit = 20;
+  static queueTimeout = 2000;
```
Comment (hypersolo) [inline ts-version/src/data/swap-executer/evm-swap-executer.ts:19]: set this to sth better (e.g. 20000)

File: ts-version/src/data/swap-executer/evm-swap-executer.ts [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+
+  public async setBroadcasters(): Promise<void> {
+    const broadcasters = await this.BlockchainNetwork.getBroadcasters();
+    EVMSwapExecuter.broadcasters.set(
+      this.BlockchainNetwork.network,
+      broadcasters as broadcaster[],
+    );
+  }
+
+  private _getBroadcaster(network: string): broadcaster | undefined {
```
Comment (hypersolo) [inline ts-version/src/data/swap-executer/evm-swap-executer.ts:47]: - check for balance
- add retry

File: ts-version/src/data/swap-executer/evm-swap-executer.ts [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+        });
+        await fillRepository.updateStatus({
+          id: fill.matchId,
+          status: FillStatus.rejected,
+        });
+        await orderRepository.commitTransaction();
+      } catch (err) {
+        await orderRepository.rollbackTransaction();
+      }
+    } finally {
```
Comment (hypersolo) [inline ts-version/src/data/swap-executer/evm-swap-executer.ts:93]: using finally block is wrong.

just call this if matchedOrder is provided and no error is thrown.

File: ts-version/src/data/swap-executer/evm-swap-executer.ts [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+            id: fill.match_id,
+            status: FillStatus.rejected,
+          }),
+          await orderRepository.commitTransaction(),
+        ]);
+      }
+      await orderRepository.rollbackTransaction();
+    } finally {
+      broadcaster.orderBroadcasting = false;
+      for (const fill of data) {
```
Comment (hypersolo) [inline ts-version/src/data/swap-executer/evm-swap-executer.ts:165]: move this for loop into try block.

Comment (hypersolo) [reply ts-version/src/data/swap-executer/evm-swap-executer.ts:165]: using finally block is wrong.

just call this if matchedOrder is provided and no error is thrown.

**Pull request merged**